### ISSUE TITLE
test_sslrecords fails sometimes

### ISSUE THREAD

**Issue by sebastianas** [openssl/openssl#1361]
Hi,

the test

`HARNESS_VERBOSE=yes make TESTS=test_sslrecords test`
fails sometimes:

> $ uniq -c results 
>    28 Okay
>     2 Failed
>     7 Okay
>     1 Failed
>     8 Okay
>    1 Failed
>     29 Okay

(28 times in row okay,  two times in row failed, …)

The output difference between "Okay" and "Failed" is
+#   Failed test 'In context empty records test'
+#   at ../test/recipes/70-test_sslrecords.t line 48.

in the failed case. The complete log for [fail](https://breakpoint.cc/openssl-test-fail) and [pass](https://breakpoint.cc/openssl-test-pass).  Any ideas what might cause it?

This is on Debian-Hurd btw. The same test passed ~270 times in a row on Debian-x86_64.

**Comment by mattcaswell** [comment]
Does it always fail in exactly the same way and at the same line?

**Comment by sebastianas** [comment]
On 2016-07-29 09:20:04 [-0700], mattcaswell wrote:
> Does it always fail in exactly the same way and at the same line?

yes. "Failed test 'In context empty records test'" is the only thing
that fails.

If you diff bad vs good, you see:

-
-#   Failed test 'In context empty records test'
-#   at ../test/recipes/70-test_sslrecords.t line 48.

…

+Received client packet
+Packet length = 57
+Processing flight 4
+ Record 1 (client -> server)
+  Content type: APPLICATION DATA
+  Version: TLS1.2
+  Length: 52
+  [ENCRYPTED APPLICATION DATA]
+  [test
+]
+
+Forwarded packet length = 57
+Received client packet
+Packet length = 57
+Processing flight 5
+ Record 1 (client -> server)
+  Content type: ALERT
+  Version: TLS1.2
+  Length: 52
+
+Forwarded packet length = 57

but no additional information of why. I can provide multiple logs of
pass and fail it you want.

Sebastian

**Comment by mattcaswell** [comment]
Please checkout and build the "test_sslrecords-debug" branch from my personal repo:
https://github.com/mattcaswell/openssl

Open 3 terminals. From the first enter:

`HARNESS_VERBOSE=1 make TESTS="test_sslrecords" test`

It should display instructions on what to enter in the other 2 terminals:

````
RUN ME FROM test DIR IN TERMINAL 2:
OPENSSL_ENGINES=../engines ../util/shlib_wrap.sh ../apps/openssl s_server -no_comp -rev -engine ossltest -accept 4443 -cert ../apps/server.pem -naccept 1 -cipher AES128-SHA
RUN ME FROM test DIR IN TERMINAL 3:
echo test | OPENSSL_ENGINES=../engines ../util/shlib_wrap.sh ../apps/openssl s_client -engine ossltest -connect localhost:4453
````

Please repeat the above steps until you see a failure, and the report back what you see in each of the three terminals. Thanks!

**Comment by sebastianas** [comment]
On 2016-07-30 03:11:46 [-0700], mattcaswell wrote:
> Please repeat the above steps until you see a failure, and the report back what you see in each of the three terminals. Thanks!

This triggers less often. So I made a "while((1)) …" loop in all three
terminals with a GO/END label including a timestamp to put them together.

Here is one extracted run:

Terminal one:

|GO: Sat Jul 30 14:42:56 CEST 2016
|make depend && make _tests
|make[1]: Entering directory '/home/bigeasy/openssl/openssl-1.1.0-pre5'
|make[1]: Leaving directory '/home/bigeasy/openssl/openssl-1.1.0-pre5'
|make[1]: Entering directory '/home/bigeasy/openssl/openssl-1.1.0-pre5'
|( cd test; \
|  SRCTOP=../. \
|  BLDTOP=../. \
|  PERL="perl" \
|  EXE_EXT= \
|  OPENSSL_ENGINES=.././engines \
|    perl .././test/run_tests.pl test_sslrecords )
|
|#   Failed test 'In context empty records test'
|#   at ../test/recipes/70-test_sslrecords.t line 48.
|# Looks like you failed 1 test of 1.
|../test/recipes/70-test_sslrecords.t .. 
|1..1
|Proxy started on port 4453
|RUN ME FROM test DIR IN TERMINAL 2:
|OPENSSL_ENGINES=../engines ../util/shlib_wrap.sh ../apps/openssl s_server -no_comp -rev -engine ossltest -accept 4443 -cert ../apps/server.pem -naccept 1 -cipher AES128-SHA
|Connection opened
|RUN ME FROM test DIR IN TERMINAL 3:
|echo test | OPENSSL_ENGINES=../engines ../util/shlib_wrap.sh ../apps/openssl s_client -engine ossltest -connect localhost:4453
|Received client packet
|Packet length = 184
|Processing flight 0
| Record 1 (client -> server)
|  Content type: HANDSHAKE
|  Version: TLS1.0
|  Length: 179
|  Message type: ClientHello
|  Message Length: 175
|    Client Version:771
|    Session ID Len:0
|    Ciphersuite len:64
|    Compression Method Len:1
|    Extensions Len:70
|
|Forwarded packet length = 189
|Received server packet
|Packet length = 1093
|Processing flight 1
| Record 1 (server -> client)
|  Content type: HANDSHAKE
|  Version: TLS1.2
|  Length: 61
|  Message type: ServerHello
|  Message Length: 57
|    Server Version:771
|    Session ID Len:0
|    Ciphersuite:47
|    Compression Method:0
|    Extensions Len:17
| Record 2 (server -> client)
|  Content type: HANDSHAKE
|  Version: TLS1.2
|  Length: 1013
|  Message type: Certificate
|  Message Length: 1009
| Record 3 (server -> client)
|  Content type: HANDSHAKE
|  Version: TLS1.2
|  Length: 4
|  Message type: ServerHelloDone
|  Message Length: 0
|
|Forwarded packet length = 1093
|Received client packet
|Packet length = 346
|Processing flight 2
| Record 1 (client -> server)
|  Content type: HANDSHAKE
|  Version: TLS1.2
|  Length: 262
|  Message type: ClientKeyExchange
|  Message Length: 258
| Record 2 (client -> server)
|  Content type: CCS
|  Version: TLS1.2
|  Length: 1
| Record 3 (client -> server)
|  Content type: HANDSHAKE
|  Version: TLS1.2
|  Length: 68
|  Message type: Finished
|  Message Length: 12
|
|Forwarded packet length = 346
|Received server packet
|Packet length = 254
|Processing flight 3
| Record 1 (server -> client)
|  Content type: HANDSHAKE
|  Version: TLS1.2
|  Length: 170
|  Message type: NewSessionTicket
|  Message Length: 166
| Record 2 (server -> client)
|  Content type: CCS
|  Version: TLS1.2
|  Length: 1
| Record 3 (server -> client)
|  Content type: HANDSHAKE
|  Version: TLS1.2
|  Length: 68
|  Message type: Finished
|  Message Length: 12
|
|Forwarded packet length = 254
|Connection closed
|Waiting for server process to close: 8545
|not ok 1 - In context empty records test
|Dubious, test returned 1 (wstat 256, 0x100)
|Failed 1/1 subtests 
|
|Test Summary Report
|-------------------
|../test/recipes/70-test_sslrecords.t (Wstat: 256 Tests: 1 Failed: 1)
|  Failed test:  1
|  Non-zero exit status: 1
|Files=1, Tests=1,  0 wallclock secs ( 0.04 usr +  0.00 sys =  0.04 CPU)
|Result: FAIL
|Failed 1/1 test programs. 1/1 subtests failed.
|Makefile:140: recipe for target '_tests' failed
|make[1]: *** [_tests] Error 1
|make[1]: Leaving directory '/home/bigeasy/openssl/openssl-1.1.0-pre5'
|Makefile:138: recipe for target 'tests' failed
|make: *** [tests] Error 2
|DONE: Sat Jul 30 14:43:02 CEST 2016

Terminal two:

|GO: Sat Jul 30 14:42:56 CEST 2016
|engine "ossltest" set.
|Using default temp DH parameters
|ACCEPT
|CONNECTION ESTABLISHED
|Protocol version: TLSv1.2
|Client cipher list: ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:DHE-RSA-AES256-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:DHE-RSA-AES128-SHA:ECDHE-ECDSA-DES-CBC3-SHA:ECDHE-RSA-DES-CBC3-SHA:DHE-RSA-DES-CBC3-SHA:AES256-GCM-SHA384:AES128-GCM-SHA256:AES256-SHA256:AES128-SHA256:AES256-SHA:AES128-SHA:DES-CBC3-SHA:SCSV
|Ciphersuite: AES128-SHA
|Signature Algorithms: RSA+SHA512:DSA+SHA512:ECDSA+SHA512:RSA+SHA384:DSA+SHA384:ECDSA+SHA384:RSA+SHA256:DSA+SHA256:ECDSA+SHA256:RSA+SHA224:DSA+SHA224:ECDSA+SHA224:RSA+SHA1:DSA+SHA1:ECDSA+SHA1
|No peer certificate
|Supported Elliptic Curve Point Formats: uncompressed:ansiX962_compressed_prime:ansiX962_compressed_char2
|Supported Elliptic Curves: X25519:P-256:P-521:P-384
|CONNECTION CLOSED
|   0 items in the session cache
|   0 client connects (SSL_connect())
|   0 client renegotiates (SSL_connect())
|   0 client connects that finished
|   1 server accepts (SSL_accept())
|   0 server renegotiates (SSL_accept())
|   1 server accepts that finished
|   0 session cache hits
|   0 session cache misses
|   0 session cache timeouts
|   0 callback cache hits
|   0 cache full overflows (128 allowed)
|DONE: Sat Jul 30 14:43:02 CEST 2016

Terminal three:

|GO: Sat Jul 30 14:42:56 CEST 2016
|engine "ossltest" set.
|depth=0 C = UK, O = OpenSSL Group, OU = FOR TESTING PURPOSES ONLY, CN = Test Server Cert
|verify error:num=20:unable to get local issuer certificate
|verify return:1
|depth=0 C = UK, O = OpenSSL Group, OU = FOR TESTING PURPOSES ONLY, CN = Test Server Cert
|verify error:num=21:unable to verify the first certificate
|verify return:1
|CONNECTED(00000003)
|---
|Certificate chain
| 0 s:/C=UK/O=OpenSSL Group/OU=FOR TESTING PURPOSES ONLY/CN=Test Server Cert
|   i:/C=UK/O=OpenSSL Group/OU=FOR TESTING PURPOSES ONLY/CN=OpenSSL Test Intermediate CA
|---
|Server certificate
|-----BEGIN CERTIFICATE-----
|MIID5zCCAs+gAwIBAgIJALnu1NlVpZ6zMA0GCSqGSIb3DQEBBQUAMHAxCzAJBgNV
|BAYTAlVLMRYwFAYDVQQKDA1PcGVuU1NMIEdyb3VwMSIwIAYDVQQLDBlGT1IgVEVT
|VElORyBQVVJQT1NFUyBPTkxZMSUwIwYDVQQDDBxPcGVuU1NMIFRlc3QgSW50ZXJt
|ZWRpYXRlIENBMB4XDTExMTIwODE0MDE0OFoXDTIxMTAxNjE0MDE0OFowZDELMAkG
|A1UEBhMCVUsxFjAUBgNVBAoMDU9wZW5TU0wgR3JvdXAxIjAgBgNVBAsMGUZPUiBU
|RVNUSU5HIFBVUlBPU0VTIE9OTFkxGTAXBgNVBAMMEFRlc3QgU2VydmVyIENlcnQw
|ggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDzhPOSNtyyRspmeuUpxfNJ
|KCLTuf7g3uQ4zu4iHOmRO5TQci+HhVlLZrHF9XqFXcIP0y4pWDbMSGuiorUmzmfi
|R7bfSdI/+qIQt8KXRH6HNG1t8ou0VSvWId5TS5Dq/er5ODUr9OaaDva7EquHIcMv
|vPQGuI+OEAcnleVCy9HVEIySrO4P3CNIicnGkwwiAud05yUAq/gPXBC1hTtmlPD7
|TVcGVSEiJdvzqqlgv02qedGrkki6GY4S7GjZxrrf7Foc2EP+51LJzwLQx3/JfrCU
|41NEWAsu/Sl0tQabXESN+zJ1pDqoZ3uHMgpQjeGiE0olr+YcsSW/tJmiU9OiAr8R
|AgMBAAGjgY8wgYwwDAYDVR0TAQH/BAIwADAOBgNVHQ8BAf8EBAMCBeAwLAYJYIZI
|AYb4QgENBB8WHU9wZW5TU0wgR2VuZXJhdGVkIENlcnRpZmljYXRlMB0GA1UdDgQW
|BBSCvM8AABPR9zklmifnr9LvIBturDAfBgNVHSMEGDAWgBQ2w2yI55X+sL3szj49
|hqshgYfa2jANBgkqhkiG9w0BAQUFAAOCAQEAqb1NV0B0/pbpK9Z4/bNjzPQLTRLK
|WnSNm/Jh5v0GEUOE/Beg7GNjNrmeNmqxAlpqWz9qoeoFZax+QBpIZYjROU3TS3fp
|yLsrnlr0CDQ5R7kCCDGa8dkXxemmpZZLbUCpW2Uoy8sAA4JjN9OtsZY7dvUXFgJ7
|vVNTRnI01ghknbtD+2SxSQd3CWF6QhcRMAzZJ1z1cbbwGDDzfvGFPzJ+Sq+zEPds
|xoVLLSetCiBc+40ZcDS5dV98h9XD7JMTQfxzA7mNGv73JoZJA6nFgj+ADSlJsY/t
|JBv+z1iQRueoh9Qeee+ZbRifPouCB8FDx+AltvHTANdAq0t/K3o+pplMVA==
|-----END CERTIFICATE-----
|subject=/C=UK/O=OpenSSL Group/OU=FOR TESTING PURPOSES ONLY/CN=Test Server Cert
|issuer=/C=UK/O=OpenSSL Group/OU=FOR TESTING PURPOSES ONLY/CN=OpenSSL Test Intermediate CA
|---
|No client certificate CA names sent
|---
|SSL handshake has read 1347 bytes and written 530 bytes
|Verification error: unable to verify the first certificate
|---
|New, SSLv3, Cipher is AES128-SHA
|Server public key is 2048 bit
|Secure Renegotiation IS supported
|Compression: NONE
|Expansion: NONE
|No ALPN negotiated
|SSL-Session:
|    Protocol  : TLSv1.2
|    Cipher    : AES128-SHA
|    Session-ID: 000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F
|    Session-ID-ctx: 
|    Master-Key: 000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F000102030405060708090A0B0C0D0E0F
|    PSK identity: None
|    PSK identity hint: None
|    SRP username: None
|    TLS session ticket lifetime hint: 7200 (seconds)
|    TLS session ticket:
|    0000 - 0a 0b 0c 0d 0e 0f 10 11-12 13 0a 0b 0c 0d 0e 0f   ................
|    0010 - 0a 0b 0c 0d 0e 0f 10 11-12 13 0a 0b 0c 0d 0e 0f   ................
|    0020 - 76 e8 e8 05 c3 14 e0 5e-b1 37 96 69 e7 8f b4 94   v......^.7.i....
|    0030 - bc e7 ef 1e 79 58 8c 6f-38 32 fe b2 25 3f 2f ab   ....yX.o82..%?/.
|    0040 - 4a 97 79 22 8e 22 b3 9f-a0 0c fb 46 dc f8 05 f9   J.y".".....F....
|    0050 - 95 c8 b3 c1 5f 7b 2d 1a-42 3e b2 d1 3f 5c 1a 3f   ...._{-.B>..?\.?
|    0060 - fe c7 41 a4 fd 0f 92 6e-1d 7b 27 ed 49 4f 1e 66   ..A....n.{'.IO.f
|    0070 - 01 c2 e0 ce df 2c 94 47-dd 02 d9 72 b7 92 81 29   .....,.G...r...)
|    0080 - 00 01 02 03 04 05 06 07-08 09 0a 0b 0c 0d 0e 0f   ................
|    0090 - 10 11 12 13 14 15 16 17-18 19 1a 1b 1c 1d 1e 1f   ................
|
|    Start Time: 1469882576
|    Timeout   : 7200 (sec)
|    Verify return code: 21 (unable to verify the first certificate)
|    Extended master secret: yes
|---
|DONE
|DONE: Sat Jul 30 14:42:56 CEST 2016

All runs:
  https://breakpoint.cc/ossl-tc-term-pass-fail/term1.txt
  https://breakpoint.cc/ossl-tc-term-pass-fail/term2.txt
  https://breakpoint.cc/ossl-tc-term-pass-fail/term3.txt

The extracted one iteration:
  https://breakpoint.cc/ossl-tc-term-pass-fail/term1-one.txt
  https://breakpoint.cc/ossl-tc-term-pass-fail/term2-one.txt
  https://breakpoint.cc/ossl-tc-term-pass-fail/term3-one.txt

Sebastian

**Comment by mattcaswell** [comment]
Hmmmm. This looks suspiciously like some problem on the s_client side. This test is setting up a communication between s_server and s_client with a perl proxy in the middle. The proxy's job in this case is to "fiddle" with some of the packets to set up various test scenarios and then check that the two ends of the communication behave as expected.

In this case s_client and s_server are supposed to successfully be able to establish a connection - and in fact they do. You can see in both the s_server and s_client logs (terminals 2 and 3 respectively) that even in the case where the test fails, they have in fact successfully established a connection. The next thing that is supposed to happen is that s_client is supposed to send some application data and then close down the connection cleanly by sending a close_notify alert. This data never gets sent, i.e. we are missing this bit when the test fails (as you noted above):

````
+Received client packet
+Packet length = 57
+Processing flight 4
+ Record 1 (client -> server)
+  Content type: APPLICATION DATA
+  Version: TLS1.2
+  Length: 52
+  [ENCRYPTED APPLICATION DATA]
+  [test
+]
+
+Forwarded packet length = 57
+Received client packet
+Packet length = 57
+Processing flight 5
+ Record 1 (client -> server)
+  Content type: ALERT
+  Version: TLS1.2
+  Length: 52
+
+Forwarded packet length = 57
````

The s_client command line looks like this:

````
echo test | OPENSSL_ENGINES=../engines ../util/shlib_wrap.sh ../apps/openssl s_client -engine ossltest -connect localhost:4453
````

The interesting bit here is the `echo test |` at the beginning. Once s_client successfully establishes it's connection it is supposed to read data from stdin and send it through to the server. In this case the string "test". Subsequently, after there is no more data to read from stdin, it is supposed to cleanly close down. For some reason that is not happening.

Given that this works on every other platform (AFAIK), I suspect some peculiarity in the way that Hurd is handling the pipe and presenting the stdin to s_client. Although why only this test is affected is a mystery to me.

As I don't have access to a Hurd system though, I'm not really sure how best to proceed.

**Comment by sebastianas** [comment]
On 2016-08-01 03:40:10 [-0700], mattcaswell wrote:
> The interesting bit here is the `echo test |` at the beginning. Once s_client successfully establishes it's connection it is supposed to read data from stdin and send it through to the server. In this case the string "test". Subsequently, after there is no more data to read from stdin, it is supposed to cleanly close down. For some reason that is not happening.

So it might be a libc / kernel issue.

> Given that this works on every other platform (AFAIK), I suspect some peculiarity in the way that Hurd is handling the pipe and presenting the stdin to s_client. Although why only this test is affected is a mystery to me.

:) Thanks for looking into this.

> As I don't have access to a Hurd system though, I'm not really sure how best to proceed.

Hurd is not a primary Debian architecture so my motivation is quite low
now that it mostly builds. I will point the Debian Hurd porter(s) to this
and maybe they dig deeper.

However if you _want_ hurd access, you probably can get one:
  https://www.gnu.org/software/hurd/public_hurd_boxen.html

and I understand if you don't.

Sebastian

**Comment by mattcaswell** [comment]
Ok. I'm going to keep this open, but move the milestone to post 1.1.0. Unless someone keen with some Hurd knowledge comes along it looks unlikely that this will get fixed in that timescale.

**Comment by levitte** [comment]
Is this still an issue?

**Comment by sebastianas** [comment]
I'm not a hurd user myself. I just digged into this as part of other hurd failures which got fixed. And from what Matt wrote here it did look like something that could openssl related. I reported this as a possible issue to the debian hurd maintainer and I have no idea if their proceeded.
If you want I could test the master branch and see how it goes. Other than that I have no objections if this gets closed because it does not look as an openssl issue.
